BaseStripeTest class with stripe-mock setup

ob-stripe opened the pull request:
r? @remi-stripe @brandur-stripe 
cc @stripe/api-libraries 

Adds a `BaseStripeTest` class that takes care of setting up and tearing down all the stripe-mock stuff.

Also adds a `StripeCouponServiceTest` to illustrate how to use `BaseStripeTest` and hopefully serve as a template for other tests.

`StripeCouponServiceTest` implements `IClassFixture<BaseStripeTest>` rather than just `BaseStripeTest` so that the initialization is done once for the entire class rather than for every test. 

Unfortunately, because the base URLs can only be set globally, I had to disable parallel test execution by grouping all tests in the same collection (by default, tests are grouped by class [0]). One alternative would be to add the ability to provide the base URLs in `StripeRequestOptions` (as can already be done for the API key) and pass an instance with the correct values in all the stripe-mock tests.

[0] https://xunit.github.io/docs/running-tests-in-parallel.html#runners-and-test-frameworks

brandur-stripe commented:
Slick OB!

> `StripeCouponServiceTest` implements `IClassFixture<BaseStripeTest>` rather than just `BaseStripeTest` so that the initialization is done once for the entire class rather than for every test. 

Nice touch.

> Unfortunately, because the base URLs can only be set globally, I had to disable parallel test execution by grouping all tests in the same collection (by default, tests are grouped by class [0]). One alternative would be to add the ability to provide the base URLs in `StripeRequestOptions` (as can already be done for the API key) and pass an instance with the correct values in all the stripe-mock tests.

IMO ,adding the ability to set this configuration per-request (or preferably per-client although I don't think we really have such a concept in .NET) would be better.

jaymedavis commented on src/Stripe.Tests.XUnit/BaseStripeTest.cs:
```diff
@@ -0,0 +1,106 @@
+            CheckStripeMock();
+            SetUpStripeMock();
+        }
+
```
One thing I liked in @remi-stripe's other PR was a destructor. Having additional cleanup area could prove useful. 

~BaseStripeTest()
{
}

jaymedavis commented on src/Stripe.Tests.XUnit/BaseStripeTest.cs:
```diff
@@ -0,0 +1,106 @@
+        public void Dispose()
+        {
+            TearDownStripeMock();
+        }
```
You'd need to suppress the garbage collection in dispose for most cases

```GC.SuppressFinalize(this);```

ob-stripe commented:
@jaymedavis Thanks for the pointers! I found a StackOverflow answer that seems relevant: https://stackoverflow.com/a/151244/5307473. I don't think we need this kind of additional complexity at this point, but I'll keep it in mind.

> IMO ,adding the ability to set this configuration per-request (or preferably per-client although I don't think we really have such a concept in .NET) would be better.

I tried taking a stab at this, but I think it will require some refactoring. Currently, services form the entire URL using the `Urls` package, which directly reads from `StripeConfiguration.GetApiBase()`. I think we should modify it to have services form the _path_ only, and have the `Requestor` class then form the full URL by joining the base URL and the path.

ob-stripe commented:
I've made a few updates to the PR:
- stripe-mock is now cached on AppVeyor's end, so that it's only downloaded if `appveyor.yml` changes (e.g. when bumping `STRIPE_MOCK_VERSION`)
- added a deserialization test in `Entities/StripeCouponTest.cs`, which uses a `GetFixture()` method similar to what we do in stripe-java
- added a (very basic) serialization test in `Services/StripeCouponCreateOptionsTest.cs`. I'm not 100% sure about this one, in particular I wonder if serialization is deterministic (i.e. if parameters will always be in the same order)
- changed the `BaseStripeTest` class to only run the stripe-mock check once (via static variables). This way tests can directly inherit `BaseStripeTest` and benefit from any helper methods we define (e.g. `GetFixture`)

remi-stripe commented on src/Stripe.Tests.XUnit/BaseStripeTest.cs:
```diff
@@ -0,0 +1,161 @@
+
+        private static string Port;
+
+        private string OrigApiBase;
```
Orig => Original ? (stripe-dotnet does not really abbreviate variable names, let's not start that trend)

remi-stripe commented on src/Stripe.Tests.XUnit/BaseStripeTest.cs:
```diff
@@ -0,0 +1,161 @@
+
+        private string OrigApiBase;
+        private string OrigUploadBase;
+        private string OrigApiKey;
```
Should be alphabetize while we're at it?

remi-stripe commented on src/Stripe.Tests.XUnit/BaseStripeTest.cs:
```diff
@@ -0,0 +1,161 @@
+        {
+            if (string.IsNullOrEmpty(Port))
+            {
+                CheckStripeMock();
```
Start with `this.` on all methods that are in the same class.

remi-stripe commented on src/Stripe.Tests.XUnit/BaseStripeTest.cs:
```diff
@@ -0,0 +1,161 @@
+
+        public BaseStripeTest()
+        {
+            if (string.IsNullOrEmpty(Port))
```
Start with `this.` for `Port` and all other class members

remi-stripe commented on src/Stripe.Tests.XUnit/BaseStripeTest.cs:
```diff
@@ -0,0 +1,161 @@
+                    );
+                }
+
+                return response.Content.ReadAsStringAsync().Result;
```
This only supports stripe-mock fixtures right? We should add a JSON one like in Java, those really are the most important ones to have/test.

ob-stripe replied:
Yep. Ideally I'd like `BaseStripeTest` to have the same helper methods as we have in stripe-java, so it's still missing things like:
- retrieving contents of a local fixture file
- stubbing requests
- verifying the requested URL / parameters

The last two will be fairly complex to implement (and will likely require using a library like [moq](https://github.com/moq/moq4)) but the first one shouldn't be too hard.

remi-stripe commented on src/Stripe.Tests.XUnit/BaseStripeTest.cs:
```diff
@@ -0,0 +1,161 @@
+                    throw new Exception(
+                        $"Your version of stripe-mock ({version}) is too old. The minimum "
+                        + $"version to run this test suite is {MockMinimumVersion}. Please see its "
+                        + "repository for upgrade instructions."
```
OOC why don't we require the exact same version? New versions will likely break tests too (that happened to be at least twice in Java)

ob-stripe replied:
Yes, but that's what you want:
1. Upgrade stripe-mock
2. See which tests fail, and fix the code (or the tests themselves)

If the version check prevents the test suite from running at all, then 2. becomes more difficult :)

remi-stripe commented on src/Stripe.Tests.XUnit/Properties/AssemblyInfo.cs:
```diff
@@ -0,0 +1,6 @@
+// By default, group all tests in a single collection. This ensures the tests are run sequentially,
+// which is necessary because we have to change the global configuration in order to use
+// stripe-mock.
+[assembly: CollectionBehavior(CollectionBehavior.CollectionPerAssembly)]
```
maybe add a todo to remove that once we have completed the migration to stripe-mock

remi-stripe commented on src/Stripe.Tests.XUnit/Services/Coupons/StripeCouponServiceTest.cs:
```diff
@@ -0,0 +1,105 @@
+        public void List()
+        {
+            var coupons = Service.List(ListOptions);
+            Assert.NotNull(coupons);
```
Should we confirm what's _in_ the list?

ob-stripe replied:
Not sure there'd be much value in it since I think whatever stripe-mock returns will be coerced into the expected type. I'll check though.

remi-stripe replied:
yeah you're likely right. I always have a hard time letting go of all the code that tested (and caught) so many thing before but it's likely superfluous and we just need good tests for `StripeList` in general

remi-stripe commented on src/Stripe.Tests.XUnit/Services/Coupons/StripeCouponCreateOptionsTest.cs:
```diff
@@ -0,0 +1,29 @@
+            };
+
+            var url = this.Service.ApplyAllParameters(options, "", false);
+            Assert.Equal("?percent_off=25&duration=forever", url);
```
I like this approach but not entirely sure we want this for every single resource. I think something like what we have in Go to test all "types" could work better.
Otherwise we likely want to only test complex types like lists? We also should test all options class otherwise so create and update here.

ob-stripe replied:
Right, to be clear, this particular test is fairly useless and I completely agree that most of this stuff should be tested only once on `ParameterBuilder`. This was just an example of what a custom serialization test could work. In practice, we'd probably only want these for classes that use custom plugins.

ob-stripe commented:
Alright, I moved the new tests into a new project `StripeTests`, and disabled `StyleCop.Analyzers` in the existing `Stripe.Tests.Xunit` project. Should make it easier to ensure that our new tests are compliant with the linting rules without getting tens of thousands of warnings.

remi-stripe commented on src/StripeTests/BaseStripeTest.cs:
```diff
@@ -0,0 +1,158 @@
+
+        private static string port;
+
+        private string originalApiBase;
```
since we are moving tests to a new package, they will all be based on stripe-mock (or in the future stubbed). Can we get rid of all this in that case? Or are the test suites ran in parallel and so they conflict?

ob-stripe replied:
I think there's _some_ value in cleaning up (though tbh we could simply restore the default values instead of saving the original values).

The tests are run in parallel on AppVeyor (not sure why they aren't on our laptops -- xUnit uses the "number of virtual CPUs in the PC" by default, I guess that xUnit only sees 1 for some reason). It should still work because all the tests in this project will use the same global state values, though ideally we shouldn't rely on this. A better solution would be to be able to provide all options directly in each tests, via `RequestOptions` or an hypothetical `StripeClient`, but that's not possible right now.

jaymedavis replied:
> xUnit uses the "number of virtual CPUs in the PC" by default
Interesting. Never knew that.

You could create a `TestRequestOptions` of sorts which has properties for api base, upload base, and api key (private strings on BaseStripeTest now). Have the constructor of BaseStripeTest take a parameter of type `TestRequestOptions`. If the test doesn't provide this object, have it use the default set.

ob-stripe replied:
That sounds like a good idea if a specific test needs to use different values, but from past experience from the other language libraries, I don't anticipate that being the case. In other libraries, tests fall into one of two categories:
- most use stripe-mock
- some use stubbed requests (and thus don't care about values for `ApiBase` etc. since they don't make network requests)

remi-stripe commented:
I think this PR is almost there. Do we want to slowly clean the relevant files/resources tested in the 2 other test suites? Or do we simply plan to migrate everything first and then later remove the other test suite in one go?

jaymedavis commented:
You could mark them `[Obselete]` as you go. Perhaps configure the test runner to skip those.

remi-stripe commented:
@jaymedavis yeah but it's even more work than just removing the files at that point

ob-stripe commented:
No strong preference. If we do decide to clean up as we go, flagging the tests as `[Obsolete]` is probably better -- earlier in the PR I tried removing all the coupons related stuff from the old test suite, only to find in CI that I broke other tests that depended on the coupons fixtures.

remi-stripe commented on src/StripeTests/BaseStripeTest.cs:
```diff
@@ -0,0 +1,158 @@
+
+        public void Dispose()
+        {
+            this.TearDownStripeMock();
```
I added extra tests and this line causes one of them to fail. It works fine without it which seems to point to some kind of parallelism issue.

ob-stripe replied:
Ah right, I guess that one test could restore the original state while another test is ongoing.

Honestly, the right answer here is to disable parallel test execution as long as we have to rely on global state. It shouldn't be too much of an issue since stripe-mock tests are fast anyway.

remi-stripe replied:
But if the test suites runs separately (like as 3 separate projects they must right?) then that new one does not need the hack since it would only ever use stripe-mock right? This was only needed before you decided to add a third project I think.

ob-stripe replied:
Yes, the different test projects are run separately. Here's what's currently happening in the new project:
- T1: call `SetUpStripeMock()`
- T1: run a test
- T2: call `SetUpStripeMock()` (effectively a no-op since the values are the same)
- T1: call `TearDownStripeMock()`
- T2: run a test --> fails because T1 restored the original state values and so T2's requests are sent to the live API servers instead of stripe-mock

We could simply set the global state once then never clean up and that would probably work, but it seems less hacky to me to disable parallel execution until we no longer have to rely on global state (i.e. until we can provide `ApiBase` / `UploadBase` on a per-request or per-service basis).

ob-stripe replied:
It just seems cleaner to me to have isolated tests that don't rely on global state.

That said, it looks like stripe-go does exactly this (i.e. sets the global configuration once and never restore it afterwards) so I've updated the PR to do the same thing.

ob-stripe commented:
ptal @remi-stripe

ob-stripe commented:
Sight, this was more annoying than I thought, but it should be working now. I've changed the initialization to use a `Lazy<T>` class. Normally, this is used to defer the initialization of some object until the first time it's needed. In this case, we're not really initializing an object (just setting some global state) so the initializer factory method (`InitStripeMock`) simply returns `null`, and we don't really care about deferring initialization so we call `initializer.Value` (which triggers the lazy initialization) in `BaseStripeTest`'s constructor.

The interesting property of `Lazy<T>` is that they're thread-safe:
> By default, `Lazy<T>` objects are thread-safe. That is, if the constructor does not specify the kind of thread safety, the `Lazy<T>` objects it creates are thread-safe. In multi-threaded scenarios, the first thread to access the `Value` property of a thread-safe `Lazy<T>` object initializes it for all subsequent accesses on all threads, and all threads share the same data. Therefore, it does not matter which thread initializes the object, and race conditions are benign.

So this should ensure that even when running tests in parallel, initialization will only be run once and threads will block until initialization is complete.

remi-stripe commented on src/StripeTests/BaseStripeTest.cs:
```diff
@@ -0,0 +1,157 @@
+        /// </summary>
+        private static void SetUpStripeMock()
+        {
+        }
```
is this still needed?

ob-stripe replied:
Nope. Good catch, thanks.

remi-stripe commented on src/StripeTests/Entities/StripeCouponTest.cs:
```diff
@@ -0,0 +1,23 @@
+            var coupon = Mapper<StripeCoupon>.MapFromJson(json);
+            Assert.NotNull(coupon);
+            Assert.IsType<StripeCoupon>(coupon);
+            Assert.NotNull(coupon.Id);
```
We discussed yesterday that we should test `Assert.Equal("coupon", coupon.Object)` here too

remi-stripe commented on src/StripeTests/Services/Coupons/StripeCouponServiceTest.cs:
```diff
@@ -0,0 +1,112 @@
+
+            this.listOptions = new StripeCouponListOptions()
+            {
+                Limit = 1,
```
Do we care to set that `Limit` if we don't explicitly test it?

ob-stripe replied:
Yeah, I think so -- just like for creation and update requests, it at least ensures that the method does accept a `StripeCouponListOptions` and that the resulting encoding is accepted by stripe-mock.

remi-stripe replied:
sure but I meant you could do this without setting `Limit`.
